src/hamiltonian.jl

```diff
@@ -138,15 +138,32 @@ phasepoint(
     h::Hamiltonian
 ) where {T<:Real} = phasepoint(h, θ, rand(rng, h.metric))
 
+abstract type AbstractRefreshment end
+
+"Completly resample new momentum."
+struct FullRefreshment <: AbstractRefreshment end
+
 refresh(
     rng::Union{AbstractRNG, AbstractVector{<:AbstractRNG}},
+    ::FullRefreshment,
```

Comment:
Hi @xukai92, as discussed earlier today, I think we can 

1. Define `Trajectory` as a subtype of `AbstractMCMCKernel`
2. Add a field `MomentumRefreshment` to `Trajectory`, or `Hamiltonian`. 

This is probably more consistent with our `Trajectory` type, which is effectively an MCMC kernel -- `Trajectory` contains how we numerically simulate a trajectory, and how to pick a phase point from a simulated trajectory. Therefore wrapping the `Trajectory` type inside another kernel type is a bit redundant in my view.

Reply:
Yes I can see this alternative has its own pros by being less nested.

Also to record the advatange of the current design: it makes introducing a new type of HMC kernel easier as one doesn't need to deal with refreshment and integrator stuff *within the trajectory*.
This is a good result of having this nested design because all pre-simulation code appears only in [this function](https://github.com/TuringLang/AdvancedHMC.jl/pull/247/files#r579826616) now, which are previously located in  the transition function for each trajecotry, seperately. I personally found this is useful from a development potin of view because I did encounter this problem when refactoring the integrator code -- I had to find all the places that the integrator was jittered.

As both sides have its own pros and cons, I guess it's better for us to discuss it a bit before making changes.

Reply:
I created a discussion here https://github.com/TuringLang/AdvancedHMC.jl/discussions/250.